sympy/signalprocessing/sp.py

```diff
@@ -0,0 +1,102 @@
+from sympy.simplify.radsimp import fraction
+from sympy.solvers import solve
+from sympy.core.numbers import oo, I
+from sympy.core.symbol import Symbol
+from sympy.functions.elementary.exponential import exp
+from sympy.functions.elementary.complexes import Abs, arg, re, im
+from sympy.utilities.lambdify import lambdify
+import math
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+# Function for obtaining the poles and zeros of a transfer function
+def pole_zero(expression, symbol):
+    num, den = fraction(expression)
+    poles_2 = solve(den, symbol)
+    zeros_2 = solve(num, symbol)
+    poles_1 = set(poles_2) - set(zeros_2)
```

Comment:
Perhaps this can be tightened up a bit:
```python
def pole_zero(expression, symbol):
    num, den = expression.as_numer_denom()
    poles = list(ordered(set(solve(den, symbol))))
    zeros = list(ordered(set(solve(num, symbol))))
    poles_extra = len(poles) - len(zeros)
  # make them the same length
    poles += [S.Zero]*-poles_extra
    zeros += [S.Infinity]*poles_extra
  # if the corresponding values are not the same
  # extend the corresponding lists with those
  # values
    for i in range(len(poles)):
        p, z = poles[i], zeros[i]
        if p != z:
            poles.append(p)
            zeros.append(z)
    return np.asarray(poles), np.asarray(zeros)
```

Why are you returning np arrays?

Is this the expected behavior:
```
>>> pole_zero((x**2-4)/(x-3)/(x-2)/(x-5),x)
(array([2, 3, 5, 2, 3, 5], dtype=object),
 array([-2, 2, oo, -2, 2, oo], dtype=object))
>>>
```

Reply:
1. The code looks tighter. However, the code does not give the appropriate output. I see, two major issues:
 a) The values of poles and zeros are repeated. ( [2,3,5 and again 2,3,5]
 b) If a pole and a zero have the same value, they should get cancelled i.e. they should not be included in both lists, poles or zeros.
 
Please refer to https://web.mit.edu/2.14/www/Handouts/PoleZero.pdf for further clarification. 

2. I honestly had no specific reason for returning numpy arrays. I just did that out of habit I guess. I could remove that of you'd like.

Reply:
Actually, the only numeric features that are accepted in SymPy are:

1. Wrapping `mpmath` numerics with `evalf`.
2. `lambdify`, printers, or code generation.
3. Internals of plotting module.

Reply:
The other library that does signal processing well is scipy.signal . It is a very extensive library that has a lot of functions that work well with actual data, but not actual mathematical expressions.
Implementing this in sympy gives the user the ability to work with actual expressions instead of just coefficients. Moreover, functionalities such as computing the expresssion of magnitude as well as frequency response are easily computed in sympy as compared to scipy.signal.

Reply:
> 1. However, the code does not give the appropriate output

I updated the code. Also, since you removed elements in common the last loop you use can't ever find a case where `zeros_1[i] == poles_1[i]`, can it?

Reply:
I also anticipate that there are a lot of things that numeric libraries cannot do, for signal processing. But I'd expect things to use more specific than using user-level solvers. (And not using numpy array to store things like `object`)
For example, if this is about solving the polynomial problem, I may consider using `Poly` class with the `roots`.

I think that things like `set(poles_2) - set(zeros_2)` can yield a lot of false-positives, because comparing the equality between two arbitrary expressions cannot be perfect.
I just wonder if it can use symbolic set complements like `FiniteSet(*poles_2) - FiniteSet(*zeros_2)` that can logically make the result correct, or if the problem statement can be modified to give roots of the numerator and the denominator polynomial.

Reply:
1. "But I'd expect things to use more specific than using user-level solvers. (And not using numpy array to store things like object)". Could you please elaborate on this, dont think i fully understand what you mean. 

2. I'll look into the Polly class and update the code accordingly. 

3. I will check whether FiniteSet does provide the same functionality and update it accordingly.

Reply:
> 
> 
> > 1. However, the code does not give the appropriate output
> 
> I updated the code. Also, since you removed elements in common the last loop you use can't ever find a case where `zeros_1[i] == poles_1[i]`, can it?

Yes that's true. Apologies. I forgot to remove the last loop before submitting the PR. I'll make the necessary changes. Thank you.

Reply:
> 
> 
> I also anticipate that there are a lot of things that numeric libraries cannot do, for signal processing. But I'd expect things to use more specific than using user-level solvers. (And not using numpy array to store things like `object`)
> For example, if this is about solving the polynomial problem, I may consider using `Poly` class with the `roots`.
> 
> I think that things like `set(poles_2) - set(zeros_2)` can yield a lot of false-positives, because comparing the equality between two arbitrary expressions cannot be perfect.
> I just wonder if it can use symbolic set complements like `FiniteSet(*poles_2) - FiniteSet(*zeros_2)` that can logically make the result correct, or if the problem statement can be modified to give roots of the numerator and the denominator polynomial.

I went through the Polly class as well as the functionalities of FiniteSet. I feel that it would over complicate the code of the current application. We could probably use them for other applications in the future.